uReplicator-Controller/src/main/java/com/uber/stream/kafka/mirrormaker/controller/core/OffsetMonitor.java

```diff
@@ -134,15 +119,45 @@ public void start() {
       int delaySec = 60 + new Random().nextInt(240);
       logger.info("OffsetMonitor starts updating offsets every {} seconds with delay {} seconds", refreshIntervalInSec,
           delaySec);
-      logger.info("OffsetMonitor starts with brokerList=" + srcBrokerList);
+
+      MetricRegistry metricRegistry = HelixKafkaMirrorMakerMetricsReporter.get().getRegistry();
+      Counter counter = metricRegistry.counter("offsetMonitor.executed");
 
       refreshExecutor.scheduleAtFixedRate(new Runnable() {
         @Override
         public void run() {
           logger.info("TopicList starts updating");
+          if (zkClientQueue == null) {
+
+            zkClientQueue = new LinkedBlockingQueue<>(numOffsetThread);
+            for (int i = 0; i < numOffsetThread; i++) {
+              ZkClient zkClient = new ZkClient(offsetZkString, 30000, 30000, ZKStringSerializer$.MODULE$);
+              zkClientQueue.add(zkClient);
+            }
+
+            ZkClient zkClient = new ZkClient(srcZkString, 30000, 30000, ZKStringSerializer$.MODULE$);
+            List<String> brokerIdList = zkClient.getChildren("/brokers/ids");
+            JSONParser parser = new JSONParser();
+
+            // Hacky solution for now to reduce init time
+            int count = 0;
+            for (String id : brokerIdList) {
+              try {
+                JSONObject json = (JSONObject) parser.parse(zkClient.readData("/brokers/ids/" + id).toString());
+                srcBrokerList.add(String.valueOf(json.get("host")) + ":" + String.valueOf(json.get("port")));
+                if (count++ >= 20) {
```

Comment:
make the threshold number configurable and default is break.

Reply:
It's not needed as we move the init part to an executor.